src/cli/NativeMsgIface.cpp

```diff
@@ -0,0 +1,454 @@
+    size_t outputSize = arr.size();
+    std::cout.write(reinterpret_cast<char*>(&outputSize), 4);
+    std::cout.write(arr.data(), outputSize);
+    std::cout.flush();
+}
+
+inline void NativeMsgIface::handleRequest(TranslationRequest myJsonInput) {
+    int myID = myJsonInput.id;
+
+    // Initialise models based on the request.
+    bool success = tryLoadModel(myJsonInput.src, myJsonInput.trg);
+    if (!success) {
+        lockAndWriteJsonHelper(errJson(myID, QString("Failed to load the necessary translation models.")));
+        operations_--;
+        return;
+    }
+
+    // Initialise translator settings options
+    marian::bergamot::ResponseOptions options;
+    options.HTML = myJsonInput.html;
+    std::function<void(marian::bergamot::Response&&)> callbackLambda = [&, myID](marian::bergamot::Response&& val) {
+        lockAndWriteJsonHelper(toJsonBytes(std::move(val), myID));
+        operations_--;};
+
+    // Attempt translation. Beware of runtime errors
+    try {
+        std::visit(overloaded {
+            [&](DirectModelInstance &model) {
+                service_->translate(model.model, std::move(myJsonInput.text.toStdString()), callbackLambda, options);
+            },
+            [&](PivotModelInstance &model) {
+                service_->pivot(model.model, model.pivot, std::move(myJsonInput.text.toStdString()), callbackLambda, options);
+            }
+        }, *model_);
+    } catch (const std::runtime_error &e) {
+       lockAndWriteJsonHelper(errJson(myID, QString::fromStdString(e.what())));
+       operations_--;
+    }
+}
+
+inline void NativeMsgIface::handleRequest(ListRequest myJsonInput)  {
+    // Fetch remote models if necessary. In this case we report the models via signal
+    if (myJsonInput.includeRemote && models_.getNewModels().isEmpty()) {
+        models_.fetchRemoteModels(myJsonInput.id);
+    } else {
+        QJsonArray modelsJson;
+        for (auto&& model : models_.getInstalledModels()) {
+            modelsJson.append(model.toJson());
+        }
+        for (auto&& model : models_.getNewModels()) {
+            modelsJson.append(model.toJson());
+        }
+        QJsonObject jsonObj {
+            {"success", true},
+            {"id", myJsonInput.id},
+            {"data", modelsJson}
+        };
+        lockAndWriteJsonHelper(QJsonDocument(jsonObj).toJson());
+        operations_--;
+    }
+
+}
+
+inline void NativeMsgIface::handleRequest(DownloadRequest myJsonInput)  {
+    // Connection to store our callback in. Qt6 has single shot connections, but 5 doesn't.
+    QMetaObject::Connection * const connection = new QMetaObject::Connection;
+
+    auto downloadModelLambda = std::function([=](){
+        // Remove the connection right after we fetch model, if the connection is valid
+        if (*connection)
+            QObject::disconnect(*connection);
+
+        // Clean up our callback connection
+        // TODO: abstract this all away in a connectSingleShot() helper?
```

Comment:
Maybe? If we do that though, we should use QT6 singleShot connections so that we can readily remove the qt5 codepath in... 5 years when the oldest ubuntu LTS ships with QT6.....